src/Core/System/Capabilities.js

```diff
@@ -17,16 +34,33 @@ export default {
         const gl = renderer.context;
         maxTexturesUnits = gl.getParameter(gl.MAX_TEXTURE_IMAGE_UNITS);
 
-        const debugInfo = gl.getExtension('WEBGL_debug_renderer_info');
-        if (debugInfo !== null) {
-            const vendor = gl.getParameter(debugInfo.UNMASKED_VENDOR_WEBGL);
-            if (vendor.indexOf('mesa') > -1 || vendor.indexOf('Mesa') > -1) {
-                maxTexturesUnits = Math.min(16, maxTexturesUnits);
+        const program = gl.createProgram();
+        const glVertexShader = _WebGLShader(renderer, gl.VERTEX_SHADER, SampleTestVS);
+
+        let fragmentShader = `#define SAMPLE ${maxTexturesUnits}\n`;
+        fragmentShader += SampleTestFS;
+
+        const glFragmentShader = _WebGLShader(renderer, gl.FRAGMENT_SHADER, fragmentShader);
+
+        gl.attachShader(program, glVertexShader);
+        gl.attachShader(program, glFragmentShader);
+        gl.linkProgram(program);
+
+        if (gl.getProgramParameter(program, gl.LINK_STATUS) === false) {
+            // eslint-disable-next-line no-console
+            console.warn(`Sampler limit exceeded: it's down from ${maxTexturesUnits} to 16`);
```

Comment:
What about:
- printing the actual error message from WebGL instead of `Sampler limit exceeded`? 
- and replacing `it's down from ${maxTexturesUnits} to 16` by `using a maximum of 16 texture units instead of the reported value (${maxTexturesUnits})`

You should also make sure that this error essage only occurs when `maxTexturesUnits`is > 16 (and add another case when it's <= 16 but the `gl.linkProgram()`fails).

Reply:
done